### PR TITLE
Improved Memory De-Allocation for CMFD

### PR DESCRIPTION
**Pull request by wbinventor**
This PR aims to improve the memory allocation and deallocation in `Cmfd`. In particular, I ran into some issues when looping over many simulations which use the same `Geometry` and `Cmfd` objects (initialized outside of some loop) but with different MGXS libraries. In particular, the MGXS libraries had different energy group structures, which has to be properly communicated to `Cmfd` by the `Solver` rather than by the `Geometry`. For example, here is a simplified version of the simulation loop:

```python
# Create OpenMOC Geometry for 17 x 17 fuel assembly
...

# Initialize CMFD
cmfd = openmoc.Cmfd()
cmfd.setSORRelaxationFactor(1.5)
cmfd.setLatticeStructure(17,17)
cmfd.setKNearest(3)
openmoc_geometry.setCmfd(cmfd)

# Loop over OpenMC MGXS Libraries with different energy group structures
for mgxs_lib in mgxs_libs:
    openmoc_materials = \
        openmoc.materialize.load_openmc_mgxs_lib(mgxs_lib, openmoc_geometry)

    # Initialize a TrackGenerator
    track_generator = openmoc.TrackGenerator(openmoc_geometry, num_azim, spacing)
    track_generator.setNumThreads(num_threads)
    track_generator.generateTracks()

    # Initialize a Solver
    solver = openmoc.CPUSolver(track_generator)
    solver.setConvergenceThreshold(tolerance)
    solver.setNumThreads(num_threads)

    # Run an eigenvalue calculation
    solver.computeEigenvalue()
    solver.printTimerReport()
```

Without this PR, one had to initialize `Cmfd` inside of the loop on each iteration. Depending on how one set up the loop, there were also memory leaks from `Cmfd` when initialized multiple times. I believe these leaks have been fixed. Although I would like to also move `TrackGenerator` initialization outside of the loop, it does not seem to be easily possible without wrestling with `Cmfd` further (since `Cmfd` gives NaN eigenvalues on all but the first iteratoin for some reason).

In addition, this PR fixes a memory leak in the `openmoc.materialize` module's methods to load MGXS libraries from HDF5 or OpenMC. The leak was only present if one loaded an MGXS library into the same `Geometry` many times, over-riding the materials in that `Geometry` each time.

### CODE REVIEW

**Comment by wbinventor** [comment]
I should note that I recognize that the revisions to `Cmfd` here do not permit user-defined coarse group structures to be specified - I'm not sure how best to add this back in but am open to ideas. Perhaps some sort of new boolean `_user_groups` class attribute for the `Cmfd` class would be helpful.

**Comment by wbinventor** [comment]
Hold off on reviewing this - I've been running into some issues "thinning" the parametric study in the code above and re-using tracks between simulations with different MGXS libraries.

File: openmoc/materialize.py [hunk 1]
```diff
@@ -131,6 +131,9 @@ def load_from_hdf5(filename='mgxs.h5', directory='mgxs',
                 # the Material must be cloned for each unique Cell
                 if material != None:
                     if len(domains) > geometry.getNumMaterials():
+                        # Return memory ownership of material to Python
+                        # so it will be deallocated after material is cloned
+                        material.thisown = 1
```
Comment (samuelshaner) [inline openmoc/materialize.py]: If I understand correctly, `thisown` can set to 0 or 1 signifying False and True. I feel like a boolean would make it a bit more obvious what is going on. What are your thoughts on setting `thisown` here (and perhaps other places) with booleans.

**Comment by wbinventor** [comment]
WHEW! Alright, this PR definitely exploded well beyond what I had planned for it, but it introduces some much needed revisions and bug fixes. Both @samuelshaner and @geogunow should look this over, esp. since our goal is to sync up both closed and open versions of the code soon. The over-arching objective behind this PR is to make it possible to more cleanly re-use objects and inter-change them between simulations. In particular, I need to run many simulations with different MGXS libraries - with different `Material` objects and/or different group structures - which is not currently possible without running into (irreproducible) memory (de-)allocation errors. This PR resolves these issues.

My original plan for this PR was to meet the following two objectives:

* Run multiple simulations with different `Material` objects within a single set of `Geometry`, `TrackGenerator`, `Cmfd` and `Solver` objects
* Run multiple simulations with different energy group structures with a single set of `Geometry`, `TrackGenerator`, `Cmfd` and `Solver` objects

Previously, I gave some example code for the types of parametric studies I want to be able to run with the same set of objects (basically the same set of `Tracks` to reduce runtime overhead). The following is a updated version of pseudocode showing the sort of "low overhead" parametric study possible with this PR:

```python
# Create OpenMOC Geometry for 17 x 17 fuel assembly
...

# Initialize CMFD
cmfd = openmoc.Cmfd()
cmfd.setSORRelaxationFactor(1.5)
cmfd.setLatticeStructure(17,17)
cmfd.setKNearest(3)
openmoc_geometry.setCmfd(cmfd)

# Initialize a TrackGenerator
track_generator = openmoc.TrackGenerator(openmoc_geometry, num_azim, spacing)
track_generator.setNumThreads(num_threads)
track_generator.generateTracks()

# Initialize a Solver
solver = openmoc.CPUSolver(track_generator)
solver.setConvergenceThreshold(tolerance)
solver.setNumThreads(num_threads)

# Loop over OpenMC MGXS Libraries with different energy group structures
for mgxs_lib in mgxs_libs:
    openmoc_materials = \
        openmoc.materialize.load_openmc_mgxs_lib(mgxs_lib, openmoc_geometry)
    solver.computeEigenvalue()
```

The diff for this PR will appear as a variety of unrelated changes to the codebase, but they all fall under the same umbrella of a philosophy for memory (de-)allocation and simulation setup. In order to permit flexibility with parametric studies, it is my opinion that we should **defer as much memory allocation and simulation setup to simulation runtime.** My goal would be to permit users to modify the material / geometric model in between simulations while safely assuming that those changes will be picked up by the `Solver` when an eigenvalue or fixed source calculation is performed. Hopefully the diff will make more sense when viewed through a lens that recognizes this was my approach here. 

On a related note, the `Geometry::initializeFlatSourceRegions()` method was refactored - and should no longer be included in inputs since the new version is called directly by the `Solver` at simulation runtime. In addition, in the process of doing my own personal tests for this PR, I updated the `GPUSolver` and `VectorizedSolver` to produce the same results for both fixed source and eigenvalue calculations as the most commonly used `CPUSolver`. The main issue (but by no means the only one) was that the formulations used to compute different types of residuals in `CPUSolver` changed sometime in the last year and the corresponding code for the `VectorizedSolver` and `GPUSolver` was left unchanged.

In conclusion, if you see a change you don't understand, just ask and I'll try to remember which of the many bugs that particular change was intended to fix ;-)

**Comment by samuelshaner** [comment]
Wow - this is fantastic! Have the ability to run parameter sweeps in the same input file will be awesome! Thanks for putting in the effort to get this working @wbinventor!

File: openmoc/materialize.py [hunk 2]
```diff
@@ -138,7 +141,11 @@ def load_from_hdf5(filename='mgxs.h5', directory='mgxs',
                     if isinstance(domain_spec, int):
                         material = openmoc.Material(id=domain_spec)
                     else:
-                        material = openmoc.Material(name=domain_spec)
+                        # Reproducibly hash the domain name into an integer ID
+                        domain_id = \
+                            int(hashlib.md5(domain_spec).hexdigest()[:4], 16)
```
Comment (samuelshaner) [inline openmoc/materialize.py]: I'm curious, what do the 4 and 16 represent here?

File: openmoc/materialize.py [hunk 3]
```diff
@@ -211,6 +220,10 @@ def load_from_hdf5(filename='mgxs.h5', directory='mgxs',
             py_printf('INFO', 'Loaded "fission" MGXS for "%s %s"',
                       domain_type, str(domain_spec))
 
+    # Inform SWIG to garbage collect any old Materials from the Geometry
+    for material_id in old_materials:
+        old_materials[material_id].thisown = False
```
Comment (samuelshaner) [inline openmoc/materialize.py]: Thanks for changing the `thisown` to be set to `True` and `False`!

File: src/CPUSolver.cpp [hunk 4]
```diff
@@ -222,16 +191,37 @@ void CPUSolver::initializeSourceArrays() {
...
+    int fsr_id, group;
+    for (iter = _fixed_sources_map.begin();
+	 iter != _fixed_sources_map.end(); ++iter) {
+      fsr_group_key = iter->first;
+      fsr_id = fsr_group_key.first;
+      group = fsr_group_key.second;
+
+      if (group <= 0 || group > _num_groups)
+        log_printf(ERROR,"Unable to use fixed source for group %d in "
+                   "in a %d energy group problem", group, _num_groups);
```
Comment (samuelshaner) [inline src/CPUSolver.cpp]: remove one of the `in`s here.

File: src/CPUSolver.cpp [hunk 5]
```diff
@@ -550,11 +540,13 @@ double CPUSolver::computeResidual(residualType res_type) {
     FP_PRECISION inverse_k_eff = 1.0 / _k_eff;
     FP_PRECISION* nu_sigma_f;
     Material* material;
+    FP_PRECISION* sigma_s;
 
     for (int r=0; r < _num_FSRs; r++) {
```
Comment (samuelshaner) [inline src/CPUSolver.cpp]: I remember Matt Everson running into some performance issues with the calculation for the scattering source taking a long time when many energy groups were used. While not directly related to this PR, seeing this for loop reminded me of that issue. Could you parallelize this loop to avoid any performance bottlenecks?

File: src/Cmfd.cpp [hunk 6]
```diff
@@ -880,6 +886,10 @@ void Cmfd::initializeMaterials() {
 
   Material* material;
 
+  /* Delete old Cmfd surface currents vector if it exists */
+  if (_materials != NULL)
+    delete _materials;
```
Comment (samuelshaner) [inline src/Cmfd.cpp]: I think this should be `delete [] _materials`, otherwise I would assume just the pointer to the array is deleted and not the actual array.

File: src/Cmfd.cpp [hunk 7]
```diff
@@ -948,11 +958,19 @@ void Cmfd::initializeCellMap() {
...
-    _group_indices_map = new int[_num_moc_groups];
+  /* Setup one-to-one fine-to-coarse group map if not specified by user */
+  if (!_user_group_indices) {
+    setGroupStructure(NULL, _num_moc_groups+1);
+    _user_group_indices = false;
   }
 
+  /* Delete old group indices map if it exists */
+  if (_group_indices_map != NULL)
+    delete _group_indices_map;
```
Comment (samuelshaner) [inline src/Cmfd.cpp]: ditto

File: src/VectorizedSolver.cpp [hunk 8]
```diff
@@ -246,6 +240,32 @@ void VectorizedSolver::initializeSourceArrays() {
...
+  /* Compute the number of SIMD vector widths needed to fit energy groups */
+  _num_vector_lengths = (_num_groups / VEC_LENGTH) + 1;
+
+  /* Reset the number of energy groups by rounding up for the number
+   * of vector widths needed to accomodate the energy groups */
+  _num_groups = _num_vector_lengths * VEC_LENGTH;
+  _polar_times_groups = _num_groups * _num_polar;
+
+  /* Allocate array of mutex locks for each FSR */
+  _FSR_locks = new omp_lock_t[_num_FSRs];
```
Comment (samuelshaner) [inline src/VectorizedSolver.cpp]: `_FSR_locks` is already allocated and initialized within `CPUSolver::initializeFSRs();` so this can be removed.

File: openmoc/materialize.py (hunk 2 above)
Comment (wbinventor) [inline openmoc/materialize.py]: This simply specifies the how many decimal places to include in the hashed ID. I'll be honest that the method *may* allow for hash collisions and redundant material IDs...

File: src/CPUSolver.cpp (hunk 4 above)
Comment (wbinventor) [inline src/CPUSolver.cpp]: got it

File: src/CPUSolver.cpp (hunk 5 above)
Comment (wbinventor) [inline src/CPUSolver.cpp]: Sure, I just parallelized all of the residual calculations by FSR. It's worth noting that the only *real* fix for lots of energy groups would be some sort of sparse data structure for scattering matrices, but that can be left for another day...

File: src/VectorizedSolver.cpp (hunk 8 above)
Comment (wbinventor) [inline src/VectorizedSolver.cpp]: Yes indeed!

File: src/Cmfd.cpp (hunk 7 above)
Comment (wbinventor) [inline src/Cmfd.cpp]: good catch

File: src/Cmfd.cpp (hunk 6 above)
Comment (wbinventor) [inline src/Cmfd.cpp]: agreed

File: src/CPUSolver.cpp [hunk 9]
```diff
@@ -501,6 +491,7 @@ double CPUSolver::computeResidual(residualType res_type) {
 
     norm = _num_FSRs;
 
+    #pragma omp for schedule(guided)
```
Comment (samuelshaner) [inline src/CPUSolver.cpp]: I think you mean `#pragma omp parallel for`

File: src/CPUSolver.cpp [hunk 10]
```diff
@@ -522,6 +513,8 @@ double CPUSolver::computeResidual(residualType res_type) {
     FP_PRECISION* nu_sigma_f;
     Material* material;
 
+    #pragma omp for private(new_fission_source, old_fission_source, \
```
Comment (samuelshaner) [inline src/CPUSolver.cpp]: ditto

File: src/CPUSolver.cpp [hunk 11]
```diff
@@ -550,11 +543,15 @@ double CPUSolver::computeResidual(residualType res_type) {
     FP_PRECISION inverse_k_eff = 1.0 / _k_eff;
     FP_PRECISION* nu_sigma_f;
     Material* material;
+    FP_PRECISION* sigma_s;
 
+    #pragma omp for private(new_total_source, old_total_source, \
```
Comment (samuelshaner) [inline src/CPUSolver.cpp]: ditto

File: src/accel/cuda/GPUSolver.cu [hunk 12]
```diff
@@ -82,6 +82,101 @@ struct isnan_test {
 
 
 /**
+ * @brief A functor to multiply	all elements in	a Thrust vector by a constant.
```
Comment (samuelshaner) [inline src/accel/cuda/GPUSolver.cu]: odd gap here

Comment (wbinventor) [inline src/accel/cuda/GPUSolver.cu]: oh weird, doesn't even show up in my emacs

File: src/CPUSolver.cpp (hunk 9 above)
Comment (wbinventor) [inline src/CPUSolver.cpp]: indeed

**Comment by wbinventor** [comment]
This is up-to-date.

**Comment by samuelshaner** [comment]
Tested out the CPUSolver, GPUSolver, and VectorizedSolver and everything looks good. It will be nice to start running some parameter sweeps without having to use multiple input files!
